threads.go

```diff
@@ -0,0 +1,357 @@
+	settings := tserv.PullOptions(opts...)
+
+	if t.PubKey(id, lid) == nil {
+		return nil, fmt.Errorf("could not find log")
+	}
+
+	if t.PrivKey(id, lid) != nil { // This is our own log
+		return t.pullLocal(ctx, id, lid, offset, settings.Limit)
+	}
+
+	// Pull from addresses
+	return t.service.pull(ctx, id, lid, offset, settings)
+}
+
+func (t *threads) Logs(id thread.ID) []thread.LogInfo {
+	lgs := make([]thread.LogInfo, 0)
+	for _, lid := range t.ThreadInfo(id).Logs {
+		lg := t.LogInfo(id, lid)
+		lgs = append(lgs, lg)
+	}
+	return lgs
+}
+
+func (t *threads) Delete(ctx context.Context, id thread.ID) error {
+	panic("implement me")
+}
+
+func (t *threads) getPrivKey() ic.PrivKey {
+	return t.host.Peerstore().PrivKey(t.host.ID())
+}
+
+func (t *threads) createLog() (info thread.LogInfo, err error) {
+	return util.CreateLog(t.host.ID())
+}
+
+func (t *threads) getOrCreateLog(id thread.ID, lid peer.ID) (info thread.LogInfo, err error) {
+	info = t.LogInfo(id, lid)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+func (t *threads) getOwnLog(id thread.ID) (info thread.LogInfo) {
+	for _, lid := range t.LogsWithKeys(id) {
+		if t.PrivKey(id, lid) != nil {
+			info = t.LogInfo(id, lid)
+			return
+		}
+	}
+	return
+}
+
+func (t *threads) getOrCreateOwnLog(id thread.ID) (info thread.LogInfo, err error) {
+	info = t.getOwnLog(id)
+	if info.PubKey != nil {
+		return
+	}
+	info, err = t.createLog()
+	if err != nil {
+		return
+	}
+	err = t.AddLog(id, info)
+	return
+}
+
+func (t *threads) createNode(ctx context.Context, body format.Node, lg thread.LogInfo, settings *tserv.AddSettings) (thread.Record, error) {
+	if settings.Key == nil {
+		var key []byte
+		logKey := t.ReadKey(settings.Thread, settings.KeyLog)
+		if logKey != nil {
+			key = logKey
+		} else {
+			key = lg.ReadKey
+		}
+		var err error
+		settings.Key, err = crypto.ParseEncryptionKey(key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	event, err := cbor.NewEvent(ctx, t.dagService, body, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	prev := cid.Undef
+	if len(lg.Heads) != 0 {
+		prev = lg.Heads[0]
+	}
+	fk, err := crypto.ParseEncryptionKey(lg.FollowKey)
+	if err != nil {
+		return nil, err
+	}
+	rec, err := cbor.NewRecord(ctx, t.dagService, event, prev, lg.PrivKey, fk)
+	if err != nil {
+		return nil, err
+	}
+
+	t.SetHead(settings.Thread, lg.ID, rec.Cid())
+
+	return rec, nil
+}
+
+func (t *threads) pullLocal(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, limit int) ([]thread.Record, error) {
+	lg := t.LogInfo(id, lid)
+	if lg.PubKey == nil {
+		return nil, fmt.Errorf("could not find log")
+	}
+	fk, err := crypto.ParseDecryptionKey(lg.FollowKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var recs []thread.Record
+	if limit == 0 {
+		return recs, nil
+	}
+	cursor := lg.Heads[0]
+	for { // @todo: Max depth
```

Comment:
This is a bit dangerous at the moment. We have to look back from HEAD in order to collect new stuff - not ideal.

Reply:
Agree!, mem usage is linear with depth! Could be made const in the worst-case if the `limit` is capped (maybe a circular-buffer?). Other O(1) mem are possible too, yep.

Reply:
nice yeah, a circular buffer would be worth shot. you really just want to keep a moving window of size `limit`. this is a drawback of the graph direction. another option could be to just return head and let the consumer do the reversal, but that means they'll have to look at the DHT to find the older nodes between offset and head. tricky.